presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java

```diff
@@ -537,14 +550,53 @@ public String getAdditionalFailureInfo(long allocated, long delta)
         String message = format("%s, Top Consumers: %s", additionalInfo, topConsumers);
 
         if (verboseExceededMemoryLimitErrorsEnabled) {
-            List<TaskMemoryReservationSummary> memoryReservationSummaries = taskContexts.values().stream()
-                    .map(TaskContext::getMemoryReservationSummary)
-                    .filter(summary -> summary.getReservation().toBytes() > 0)
-                    .sorted(comparing(TaskMemoryReservationSummary::getReservation).reversed())
-                    .limit(3)
-                    .collect(toImmutableList());
+            List<TaskMemoryReservationSummary> memoryReservationSummaries = getTaskMemoryReservationSummaries();
             message += ", Details: " + memoryReservationSummaryJsonCodec.toJson(memoryReservationSummaries);
         }
         return message;
     }
+
+    @GuardedBy("this")
+    private List<TaskMemoryReservationSummary> getTaskMemoryReservationSummaries()
+    {
+        List<TaskMemoryReservationSummary> memoryReservationSummaries = taskContexts.values().stream()
+                .map(TaskContext::getMemoryReservationSummary)
+                .filter(summary -> summary.getReservation().toBytes() > 0)
+                .sorted(comparing(TaskMemoryReservationSummary::getReservation).reversed())
+                .limit(3)
+                .collect(toImmutableList());
+        return memoryReservationSummaries;
+    }
+
+    @GuardedBy("this")
+    private boolean hasLowPartitionCount()
+    {
+        List<TaskMemoryReservationSummary> memoryReservationSummaries = getTaskMemoryReservationSummaries();
+        if (memoryReservationSummaries == null || memoryReservationSummaries.isEmpty()) {
+            return false;
+        }
+        TaskMemoryReservationSummary topTask = memoryReservationSummaries.get(0);
+        List<OperatorMemoryReservationSummary> hashBuilderAllocations = topTask
```

Comment:
This can be extended to include any operator which can benefit from higher hash partition count. To keep it limited for scope & testing, we are targeting hash builder first.